File: dotnet/Lexy.Poc/Lexy.Poc.Core/Language/EnumDefinition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lexy.Poc.Core.Parser;

namespace Lexy.Poc.Core.Language
{
    public class EnumDefinition : RootNode
    {
        public EnumName Name { get; }

        public override string NodeName => Name.Value;

        public IList<EnumMember> Members { get; } = new List<EnumMember>();

        private EnumDefinition(string name, SourceReference reference) : base(reference)
        {
            Name = new EnumName(reference);
            Name.ParseName(name);
        }

        internal static EnumDefinition Parse(NodeName name, SourceReference reference)
        {
            return new EnumDefinition(name.Name, reference);
        }

        public override IParsableNode Parse(IParserContext context)
        {
            var line = context.CurrentLine;
            if (line.IsEmpty()) return this;

            if (line.IsComment())
            {
                throw new InvalidOperationException("No comments expected. Comment should be parsed by Document only.");
            }

            var member = EnumMember.Parse(context);
            if (member != null)
            {
                Members.Add(member);
            }
            return this;
        }

        protected override IEnumerable<INode> GetChildren()
        {
            yield return Name;

            foreach (var member in Members)
            {
                yield return member;
            }
        }

        protected override void Validate(IValidationContext context)
        {
            DuplicateChecker.Validate(
                context,
                member => member.Reference,
                member => member.Name,
                member => $"Enum member name should be unique. Duplicate name: '{member.Name}'",
                Members);
        }

        public bool ContainsMember(string name)
        {
            return Members.Any(member => member.Name == name);
        }
    }
}

File: typescipt/src/language/enums/enumDefinition.ts


export class EnumDefinition extends RootNode {
   public EnumName Name

   public override string NodeName => Name.Value;

   public Array<EnumMember> Members = list<EnumMember>(): new;

   private EnumDefinition(string name, SourceReference reference) {
     super(reference);
     Name = new EnumName(reference);
     Name.ParseName(name);
   }

   internal static parse(name: NodeName, reference: SourceReference): EnumDefinition {
     return new EnumDefinition(name.Name, reference);
   }

   public override parse(context: IParseLineContext): IParsableNode {
     let lastIndex = Members.LastOrDefault()?.NumberValue ?? -1;
     let member = EnumMember.parse(context, lastIndex);
     if (member != null) Members.Add(member);
     return this;
   }

   public override getChildren(): Array<INode> {
     yield return Name;

     foreach (let member in Members) yield return member;
   }

   protected override validate(context: IValidationContext): void {
     if (Members.Count == 0) {
       context.logger.fail(this.reference, `Enum has no members defined.`);
       return;
     }

     DuplicateChecker.validate(
       context,
       member => member.reference,
       member => member.Name,
       member => $`Enum member name should be unique. Duplicate name: '{member.Name}'`,
       Members);
   }

   public containsMember(name: string): boolean {
     return Members.Any(member => member.Name == name);
   }
}
